### PR TITLE
FOUNDATIONS: Enable SQLite & MySql Support 

### LINKED ISSUE #24: Consider using other Database Providers for Exceptions
I was thinking somewhere along the lines of using generics or some switch statement that takes in the value and based on that it uses the specific database provider(mssql,mysql,postgres,sqlite etc) with the specific sqlexception handling. Any help would be appreciate it

### LINKED ISSUE #31: Exception: Unable to cast object of type 'MySql.Data.MySqlClient.MySqlException' to type 'Microsoft.Data.SqlClient.SqlException'
Hello,
I am using this great library but for a MySql connection, and I can only get this exception: 
**Unable to cast object of type `MySql.Data.MySqlClient.MySqlException` to type `Microsoft.Data.SqlClient.SqlException`**

By checken the code, i found this:

```c#
       public void ThrowMeaningfulException(DbUpdateException dbUpdateException)
        {
            ValidateInnerException(dbUpdateException);
            SqlException sqlException = GetSqlException(dbUpdateException.InnerException);
            int sqlErrorCode = this.sqlErrorBroker.GetSqlErrorCode(sqlException);
            ConvertAndThrowMeaningfulException(sqlErrorCode, sqlException.Message);

            throw dbUpdateException;
        }
```
I guess the library should be inhanced to support also `MySqlException` at least. 

I forked the repo and fixed it, should I create a pull request ?

Thanks

### PR DESCRIPTION
**Pull request by mabroukmahdhi**
The solution should be now looking like this:

image

Closes #24 
Closes #31

### CODE REVIEW

File: EFxceptions.Core/DBContextBase.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+namespace EFxceptions.Core
+{
+    public abstract class DbContextBase<TDbException> : DbContext
+        where TDbException : DbException
+    {
+        private IEFxceptionService eFxceptionService;
+        private IDbErrorBroker<TDbException> errorBroker;
+
+        protected DbContextBase()
+           => InitializeInternalServices();
```
Comment (hassanhabib) [inline EFxceptions.Core/DBContextBase.cs:24]: Put the fat-arrow up

File: EFxceptions.Core/DBContextBase.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+
```
Comment (hassanhabib) [inline EFxceptions.Core/DBContextBase.cs:26]: Remove extra line

File: EFxceptions.Core/DBContextBase.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+// Licensed under the MIT License.
+// See License.txt in the project root for license information.
+// ---------------------------------------------------------------
+
+using EFxceptions.Brokers.DbErrors;
+using EFxceptions.Services;
+using Microsoft.EntityFrameworkCore;
+using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
```
Comment (ElbekDeveloper) [inline EFxceptions.Core/DBContextBase.cs:13]: Configure VS to sort `System` directives first and run cleanup

File: EFxceptions.Identity.Core/IdentityDbContextBase.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        where TDbException : DbException
+    {
+        protected IdentityDbContextBase()
+        { }
+
+        public IdentityDbContextBase(DbContextOptions options) : base(options)
+        { }
+    }
+
+    public abstract class IdentityDbContextBase<TUser, TRole, TKey, TUserClaim, TUserRole, TUserLogin, TRoleClaim, TUserToken, TDbException> : IdentityDbContext<TUser, TRole, TKey, TUserClaim, TUserRole, TUserLogin, TRoleClaim, TUserToken>
```
Comment (ElbekDeveloper) [inline EFxceptions.Identity.Core/IdentityDbContextBase.cs:45]: Too long

File: EFxceptions.Identity.Core/IdentityDbContextBase.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        private IEFxceptionService eFxceptionService;
+        private IDbErrorBroker<TDbException> errorBroker;
+
+        protected IdentityDbContextBase() =>
+            InitializeInternalServices();
+
+        public IdentityDbContextBase(DbContextOptions options) : base(options) =>
+            InitializeInternalServices();
+
+        private void InitializeInternalServices()
```
Comment (ElbekDeveloper) [inline EFxceptions.Identity.Core/IdentityDbContextBase.cs:65]: private methods should go after public ones

File: EFxceptions.Identity.MySql/EFxceptionsIdentityContext.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
+﻿// ---------------------------------------------------------------
+// Copyright (c) H. Habib, A. Luo, S. Weitzhandler & M. Mahdhi
+// All rights reserved.
+// Licensed under the MIT License.
+// See License.txt in the project root for license information.
+// ---------------------------------------------------------------
+
+using EFxceptions.Brokers.DbErrors;
```
Comment (ElbekDeveloper) [inline EFxceptions.Identity.MySql/EFxceptionsIdentityContext.cs:8]: Run cleanup

File: EFxceptions.Identity.MySql/EFxceptionsIdentityContext.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+       where TKey : IEquatable<TKey>
+    {
+        protected EFxceptionsIdentityContext()
+        { }
+
+        public EFxceptionsIdentityContext(DbContextOptions options) : base(options)
+        { }
+
+        protected override IDbErrorBroker<MySqlException> CreateErrorBroker() =>
+             new MySqlErrorBroker();
```
Comment (ElbekDeveloper) [inline EFxceptions.Identity.MySql/EFxceptionsIdentityContext.cs:33]: Fix indentation, please

File: EFxceptions.Identity.SQLite/Brokers/DbErrors/ISQLiteErrorBroker.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+// ---------------------------------------------------------------
+
+using EFxceptions.Brokers.DbErrors;
+using Microsoft.Data.Sqlite;
+
+namespace EFxceptions.Identity.SQLite.Brokers.DbErrors
+{
+    public interface ISQLiteErrorBroker : IDbErrorBroker<SqliteException>
+    {
+    }
```
Comment (ElbekDeveloper) [inline EFxceptions.Identity.SQLite/Brokers/DbErrors/ISQLiteErrorBroker.cs:15]: Slap these together

File: EFxceptions.Identity.SQLite/EFxceptionsIdentityContext.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        : IdentityDbContextBase<TUser, TRole, TKey, IdentityUserClaim<TKey>, IdentityUserRole<TKey>,
+            IdentityUserLogin<TKey>, IdentityRoleClaim<TKey>, IdentityUserToken<TKey>, SqliteException>
+       where TUser : IdentityUser<TKey>
+       where TRole : IdentityRole<TKey>
+       where TKey : IEquatable<TKey>
+    {
+        protected EFxceptionsIdentityContext()
+        { }
+
+        public EFxceptionsIdentityContext(DbContextOptions options) : base(options)
```
Comment (ElbekDeveloper) [inline EFxceptions.Identity.SQLite/EFxceptionsIdentityContext.cs:29]: public methods/constructors go first

File: EFxceptions.Identity.SQLite/EFxceptionsIdentityContext.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        { }
+
+        protected override IDbErrorBroker<SqliteException> CreateErrorBroker() =>
+             new SQLiteErrorBroker();
```
Comment (ElbekDeveloper) [inline EFxceptions.Identity.SQLite/EFxceptionsIdentityContext.cs:33]: extra space here

**Comment by ElbekDeveloper** [comment]
LGTM

**Pull request merged**